contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import { UUPSUpgradeable } from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import { AdapterInterface } from "./interfaces/AdapterInterface.sol";
+
+/**
+ * @title ForwarderBase
+ * @notice This contract expects to receive messages and tokens from an authorized sender on a previous layer and forwards messages and tokens
+ * to contracts on subsequent layers. Messages are intended to originate from the hub pool or other forwarder contracts.
+ * @dev This contract rejects messages which do not originate from the cross domain admin, which should be set to these contracts, depending on the
+ * sender on the previous layer. Additionally, this contract only knows information about networks which roll up their state to the network on which
+ * this contract is deployed. Messages continue to be forwarded across an indefinite amount of layers by recursively wrapping and sending messages
+ * over canonical bridges, only stopping when the following network contains the intended target contract.
+ * @dev Since these contracts use destination addresses to derive a route to subsequent networks, it requires that all of these forwarder contracts
+ * AND spoke pool contracts which require forwarders to receive messages to not have same addresses.
+ * @dev Base contract designed to be deployed on a layer > L1 to re-route messages to a further layer. If a message is sent to this contract which
+ * came from the cross domain admin, then it is routed to the next layer via the canonical messaging bridge. Tokens sent from L1 are accompanied by a
+ * message directing this contract on the next steps of the token relay.
+ * @custom:security-contact bugs@across.to
+ */
+abstract contract ForwarderBase is UUPSUpgradeable, AdapterInterface {
+    /**
+     * @notice Struct containing all information needed in order to send a message to the next layer.
+     * @dev We need to know three different addresses:
+     *   - The address of the adapter to delegatecall, so that we may propagate message and token relays.
+     *   - The address of the next layer's forwarder, so that we can continue the message and token relay chain.
+     *   - The address of the next layer's spoke pool, so that we can determine whether the following network contains the target spoke pool.
+     */
+    struct RouteIndex {
+        address adapter;
+        address forwarder;
+        address spokePool;
+    }
+
+    // Address of the contract which can relay messages to the subsequent forwarders/spoke pools and update this proxy contract.
+    address public crossDomainAdmin;
+    // Map from a destination address to the next token/message bridge route to reach the destination address. The destination
+    // address can either be a spoke pool or a forwarder. This mapping requires that no destination addresses (i.e. spoke pool or
+    // forwarder addresses) collide.
```

Comment:
```suggestion
    // forwarder addresses) collide. We can't use a chain ID as the key here because we don't have this information
    // from the call into `relayMessage` or `relayTokens`, which is where we'll load this data.
```

Reply:
https://github.com/across-protocol/contracts/pull/609/commits/d33099343f18b48c9db005412529ced8980b7c86